File: tools/migrate_lang_constants_to_array.php
<?php
// migrate_lang_constants_to_array.php
// Usage: php migrate_lang_constants_to_array.php assets/lang/en.php assets/lang/en_array.php

if ($argc < 3) {
    echo "Usage: php migrate_lang_constants_to_array.php <input_file.php> <output_file.php>\n";
    exit(1);
}

$input = $argv[1];
$output = $argv[2];

if (!file_exists($input)) {
    echo "Input file not found: $input\n";
    exit(1);
}

$lines = file($input);
$translations = [];

foreach ($lines as $line) {
    if (preg_match("/define\s*\(\s*'([^']+)'\s*,\s*'((?:[^'\\\\]|\\\\.)*)'\s*\)\s*;/", $line, $matches)) {
        $key = $matches[1];
        $value = stripcslashes($matches[2]);
        $translations[$key] = $value;
    }
}

// Optional: also try to extract array at the end (if exists)
$inArray = false;
foreach ($lines as $line) {
    if (preg_match('/return\s*\[\s*/', $line)) {
        $inArray = true;
        continue;
    }
    if ($inArray) {
        if (preg_match('/\];/', $line)) {
            $inArray = false;
            continue;
        }
        if (preg_match("/'([^']+)'\s*=>\s*'((?:[^'\\\\]|\\\\.)*)'\s*,?/", $line, $matches)) {
            $translations[$matches[1]] = stripcslashes($matches[2]);
        }
    }
}

// Write new array-based language file
$out = "<?php\nreturn [\n";
foreach ($translations as $k => $v) {
    $out .= "    '$k' => '" . addslashes($v) . "',\n";
}
$out .= "];\n";

file_put_contents($output, $out);

echo "Migration complete! New file: $output\n";

?>

File: tools/migrate_lang_constants_to_array.py
# migrate_lang_constants_to_array.py
# Usage: python migrate_lang_constants_to_array.py input_lang.php output_lang.php

import sys
import re

if len(sys.argv) < 3:
    print("Usage: python migrate_lang_constants_to_array.py input_lang.php output_lang.php")
    sys.exit(1)

input_file = sys.argv[1]
output_file = sys.argv[2]

translations = {}

with open(input_file, encoding='utf-8') as f:
    lines = f.readlines()

# Extract define('key','value');
define_re = re.compile(r"define\s*\(\s*'([^']+)'\s*,\s*'((?:[^'\\]|\\.)*)'\s*\)\s*;")
for line in lines:
    m = define_re.search(line)
    if m:
        key = m.group(1)
        value = m.group(2).encode('utf-8').decode('unicode_escape')
        translations[key] = value

# Extract array at the end (if exists)
in_array = False
array_re = re.compile(r"'([^']+)'\s*=>\s*'((?:[^'\\]|\\.)*)'")
for line in lines:
    if re.search(r'return\s*\[', line):
        in_array = True
        continue
    if in_array:
        if re.search(r'\];', line):
            in_array = False
            continue
        m = array_re.search(line)
        if m:
            key = m.group(1)
            value = m.group(2).encode('utf-8').decode('unicode_escape')
            translations[key] = value

# Write new PHP array file
with open(output_file, 'w', encoding='utf-8') as f:
    f.write("<?php\nreturn [\n")
    for k, v in translations.items():
        # Escape single quotes
        v = v.replace("'", "\\'")
        f.write(f"    '{k}' => '{v}',\n")
    f.write("];\n")

print(f"Migration complete! New file: {output_file}")
